### PR TITLE
Support UNION BY NAME set operation

### PR DESCRIPTION
**Pull request by lokax**
### UNION BY NAME
A union by column name instead of column position. #4320 
Doesn't support the same name appearing in the same select list.
```SQL
SELECT x, x FROM t1 UNION BY NAME SELECT y FROM t2
binder error
```

```SQL
SELECT test.x, x FROM test UNION BY NAME SELECT y FROM test;
binder error
```
The result name of SetOperationNode is [x, x, y], so two `x` cause  binder to throw exception.

When the exprs has alias, UNION BY NAME will use alias for comparison instead of permanent column name.
```SQL
SELECT x as a, x as b FROM t1 UNION BY NAME SELECT y FROM t2;
success
```
In order to be able to generate correct `alias_map` and `expression_map`, `reorder_idxs` is recorded.
```SQL
(SELECT x FROM t1 UNION ALL SELECT y FROM t1) UNION BY NAME SELECT 5 ORDER BY y;
success
(SELECT x FROM t1 UNION ALL SELECT y FROM t1) UNION BY NAME (SELECT z FROM t2 UNION ALL SELECT y FROM t2) ORDER BY y;
Error: Binder Error: Ambiguous name in ORDER BY!
```

### OrderBinder
```SQL
SELECT x FROM test ORDER BY # 2; (throw internal exception)
```
I added a conditional judgment to avoid this problem.
```SQL
EXPLAIN (SELECT x FROM test) UNION ALL (SELECT x FROM test) ORDER BY 1;
┌───────────────────────────┐                             
│          ORDER_BY         │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│           1 ASC           │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│           UNION           ├──────────────┐              
└─────────────┬─────────────┘ 
```
1 is incorrect, i think it should be # 0. # 0 might look friendlier. So when the expr has no alias, just use BoundColumnRefExpr's name.

### CODE REVIEW

File: src/planner/binder/query_node/bind_setop_node.cpp [hunk 1]
```diff
@@ -71,47 +95,166 @@ unique_ptr<BoundQueryNode> Binder::BindNode(SetOperationNode &statement) {
...
 	}
 
-	// figure out the types of the setop result by picking the max of both
-	for (idx_t i = 0; i < result->left->types.size(); i++) {
-		auto result_type = LogicalType::MaxLogicalType(result->left->types[i], result->right->types[i]);
-		if (!can_contain_nulls) {
-			if (ExpressionBinder::ContainsNullType(result_type)) {
-				result_type = ExpressionBinder::ExchangeNullType(result_type);
+	if (result->setop_type == SetOperationType::UNION_BY_NAME) {
+		case_insensitive_map_t<idx_t> left_names_map;
```
Comment (Mytherin) [inline src/planner/binder/query_node/bind_setop_node.cpp:116]: This function is getting very long, could we try moving this into separate functions?

File: src/parser/transform/statement/transform_select_node.cpp [hunk 2]
```diff
@@ -101,6 +102,10 @@ unique_ptr<QueryNode> Transformer::TransformSelectNode(duckdb_libpgquery::PGSele
 		case duckdb_libpgquery::PG_SETOP_INTERSECT:
 			result->setop_type = SetOperationType::INTERSECT;
 			break;
+		case duckdb_libpgquery::PG_SETOP_UNION_BY_NAME:
+			select_distinct = false;
```
Comment (Mytherin) [inline src/parser/transform/statement/transform_select_node.cpp:106]: Should we perhaps allow `UNION BY NAME` and `UNION ALL BY NAME`, to be more consistent w.r.t. the standard SQL syntax?

File: src/planner/binder/query_node/bind_setop_node.cpp [hunk 3]
```diff
@@ -71,47 +95,166 @@ unique_ptr<BoundQueryNode> Binder::BindNode(SetOperationNode &statement) {
...
+				result->right_reorder_exprs.push_back(move(right_reorder_expr));
+			}
+		}
+
+		if (!statement.modifiers.empty()) {
+			// handle the ORDER BY/DISTINCT clauses
+
+			// we recursively visit the children of this node to extract aliases and expressions that can be referenced
+			// in the ORDER BY
+			case_insensitive_map_t<idx_t> alias_map;
```
Comment (Mytherin) [inline src/planner/binder/query_node/bind_setop_node.cpp:216]: This looks mostly duplicated from below - can we try to unify the code here?

File: src/planner/binder/query_node/bind_setop_node.cpp [hunk 4]
```diff
@@ -71,47 +95,166 @@ unique_ptr<BoundQueryNode> Binder::BindNode(SetOperationNode &statement) {
...
+		for (idx_t i = 0; i < new_size; ++i) {
+			auto left_index = left_names_map.find(result->names[i]);
+			auto right_index = right_names_map.find(result->names[i]);
+			bool left_exist = left_index != left_names_map.end();
+			bool right_exist = right_index != right_names_map.end();
+			LogicalType result_type;
+			if (left_exist && right_exist) {
+				result_type = LogicalType::MaxLogicalType(left_node->types[left_index->second],
+				                                          right_node->types[right_index->second]);
+				need_reorder = need_reorder || left_index->second != i || right_index->second != i;
```
Comment (Mytherin) [inline src/planner/binder/query_node/bind_setop_node.cpp:153]: Predication here is less readable and not necessary, can we just change this into an `if` statement?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments:

File: src/parser/transform/statement/transform_select_node.cpp (hunk 2 above)
Comment (lokax) [reply src/parser/transform/statement/transform_select_node.cpp]: finished

File: src/planner/binder/query_node/bind_setop_node.cpp (hunk 1 above)
Comment (lokax) [reply src/planner/binder/query_node/bind_setop_node.cpp]: I just move this code segment to another big function and add some short comment.

**Review by taniabogatsch — APPROVED** [review]
Thanks for the PR, looks good to me (@hannes).

File: third_party/libpg_query/src_backend_parser_scan.cpp [hunk 5]
```diff
@@ -487,7 +486,7 @@ struct yy_buffer_state
 	/* Number of characters read into yy_ch_buf, not including EOB
 	 * characters.
 	 */
-	yy_size_t yy_n_chars;
+	int yy_n_chars;
```
Comment (hannes) [inline third_party/libpg_query/src_backend_parser_scan.cpp:489]: I wonder where this and other changes came from? What made this change to an int?

Comment (taniabogatsch) [reply third_party/libpg_query/src_backend_parser_scan.cpp:489]: I believe that this happened when re-generating the parser code after changes to `select.y` and `create.y`. It normally looks much worse, like in my PR for the list lambdas where I did bigger changes to the parser (https://github.com/duckdb/duckdb/pull/3694/files).

Comment (hannes) [reply third_party/libpg_query/src_backend_parser_scan.cpp:489]: I am aware this is generated, maybe this is down to generator versions

Comment (taniabogatsch) [reply third_party/libpg_query/src_backend_parser_scan.cpp:489]: I did generate the files myself now and the `int` is indeed caused by an older(?) bison version.
```
tania@laptop duckdb % /usr/bin/bison --version
bison (GNU Bison) 2.3
```
@lokax I will open a PR to your branch with the newly generated files.

Comment (lokax) [reply third_party/libpg_query/src_backend_parser_scan.cpp]: > I did generate the files myself now and the `int` is indeed caused by an older(?) bison version.
> 
> ```
> tania@laptop duckdb % /usr/bin/bison --version
> bison (GNU Bison) 2.3
> ```
> 
> @lokax I will open a PR to your branch with the newly generated files.

Thanks, but my version seems to be newer?.
```text
bison (GNU Bison) 3.5.1
```

Comment (taniabogatsch) [reply third_party/libpg_query/src_backend_parser_scan.cpp]: Yes, I was suspecting that. It is weird that a newer version uses `int` though...

**Comment by hannes** [comment]
Can you merge with master again please?

**Comment by hannes** [comment]
Thanks, can you send me an email at hannes@duckdblabs.com please? We'd like to chat.

**Pull request merged**